File: app/src-tauri/src/db/srs.rs
use chrono::{DateTime, TimeZone, Utc};
use rs_fsrs::{Card, Rating, State, FSRS};
use rusqlite::{params, OptionalExtension};
use serde::Serialize;
use tauri::State as TauriState;

use crate::db;
use crate::AppState;

/// Words introduced per db_get_due_cards call when the reviewer isn't
/// working through a review backlog — keeps a big vocabulary from dumping
/// hundreds of "new" cards into one session.
const DEFAULT_NEW_LIMIT: i64 = 20;

#[derive(Serialize)]
pub struct DueCard {
    pub word_id: i64,
    pub word: String,
    pub zh: String,
    pub level: Option<String>,
    pub context_sentence: String,
    /// "new" | "learning" | "review" | "relearning"
    pub state: String,
}

fn state_to_str(state: State) -> &'static str {
    match state {
        State::New => "new",
        State::Learning => "learning",
        State::Review => "review",
        State::Relearning => "relearning",
    }
}

fn state_from_i64(v: i64) -> State {
    match v {
        1 => State::Learning,
        2 => State::Review,
        3 => State::Relearning,
        _ => State::New,
    }
}

fn dt_to_sql(dt: DateTime<Utc>) -> String {
    dt.to_rfc3339()
}

fn sql_to_dt(s: &str) -> DateTime<Utc> {
    // Stored as RFC3339 (from dt_to_sql) or SQLite's default datetime('now')
    // format ("YYYY-MM-DD HH:MM:SS") for rows never touched by FSRS.
    DateTime::parse_from_rfc3339(s)
        .map(|d| d.with_timezone(&Utc))
        .or_else(|_| {
            chrono::NaiveDateTime::parse_from_str(s, "%Y-%m-%d %H:%M:%S")
                .map(|naive| Utc.from_utc_datetime(&naive))
        })
        .unwrap_or_else(|_| Utc::now())
}

/// Count of cards ready for review right now: the due backlog plus a capped
/// batch of never-reviewed words, using the same cap as db_get_due_cards'
/// default so the Dashboard badge matches what opening the reviewer shows.
///
/// Compares dates in Rust rather than SQL because next_review_at may hold
/// either RFC3339 (written by db_review_card) or SQLite's own datetime()
/// format (written by the older db_save_quiz_result path) — the two don't
/// sort correctly against each other as raw strings.
#[tauri::command]
pub fn db_get_review_count(conn: TauriState<'_, AppState>) -> Result<i64, String> {
    let db = db::lock_db(&conn)?;
    let now = Utc::now();

    let due_dates: Vec<String> = {
        let mut stmt = db
            .prepare("SELECT next_review_at FROM srs_records WHERE entity_type = 'word' AND next_review_at IS NOT NULL")
            .map_err(|e| e.to_string())?;
        let rows = stmt
            .query_map([], |row| row.get::<_, String>(0))
            .map_err(|e| e.to_string())?;
        let mut out = vec![];
        for row in rows {
            out.push(row.map_err(|e| e.to_string())?);
        }
        out
    };
    let due_count = due_dates.iter().filter(|s| sql_to_dt(s) <= now).count() as i64;

    let new_count: i64 = db
        .query_row(
            "SELECT COUNT(*) FROM words w LEFT JOIN srs_records sr ON sr.entity_id = w.id AND sr.entity_type = 'word' WHERE sr.id IS NULL",
            [],
            |row| row.get(0),
        )
        .unwrap_or(0);

    Ok(due_count + new_count.min(DEFAULT_NEW_LIMIT))
}

#[tauri::command]
pub fn db_get_due_cards(
    new_limit: Option<i64>,
    conn: TauriState<'_, AppState>,
) -> Result<Vec<DueCard>, String> {
    let db = db::lock_db(&conn)?;
    let new_limit = new_limit.unwrap_or(DEFAULT_NEW_LIMIT);

    let zh_expr = "COALESCE((SELECT wd.zh FROM word_definitions wd WHERE wd.word_id = w.id ORDER BY wd.sort_order LIMIT 1), '')";
    let context_expr = "COALESCE(
        (SELECT ei.context_sentence FROM extracted_items ei
         WHERE ei.kind = 'word' AND lower(ei.text) = lower(w.word) AND ei.context_sentence != ''
         ORDER BY ei.id DESC LIMIT 1),
        (SELECT wd.example_en FROM word_definitions wd
         WHERE wd.word_id = w.id AND wd.example_en IS NOT NULL AND wd.example_en != ''
         ORDER BY wd.sort_order LIMIT 1),
        ''
    )";

    let mut result = vec![];

    // Backlog: previously-scheduled reviews that are now due.
    let due_sql = format!(
        "SELECT w.id, w.word, {zh_expr}, w.level, {context_expr}, sr.state
         FROM words w
         JOIN srs_records sr ON sr.entity_id = w.id AND sr.entity_type = 'word'
         WHERE sr.next_review_at <= ?1
         ORDER BY sr.next_review_at ASC"
    );
    let now_str = dt_to_sql(Utc::now());
    {
        let mut stmt = db.prepare(&due_sql).map_err(|e| e.to_string())?;
        let rows = stmt
            .query_map(params![now_str], |row| {
                let state_i: i64 = row.get(5)?;
                Ok(DueCard {
                    word_id: row.get(0)?,
                    word: row.get(1)?,
                    zh: row.get(2)?,
                    level: row.get(3)?,
                    context_sentence: row.get(4)?,
                    state: state_to_str(state_from_i64(state_i)).to_string(),
                })
            })
            .map_err(|e| e.to_string())?;
        for row in rows {
            result.push(row.map_err(|e| e.to_string())?);
        }
    }

    // New: words never reviewed, capped so a big vocabulary doesn't flood the session.
    let new_sql = format!(
        "SELECT w.id, w.word, {zh_expr}, w.level, {context_expr}
         FROM words w
         LEFT JOIN srs_records sr ON sr.entity_id = w.id AND sr.entity_type = 'word'
         WHERE sr.id IS NULL
         ORDER BY w.created_at ASC
         LIMIT ?1"
    );
    {
        let mut stmt = db.prepare(&new_sql).map_err(|e| e.to_string())?;
        let rows = stmt
            .query_map(params![new_limit], |row| {
                Ok(DueCard {
                    word_id: row.get(0)?,
                    word: row.get(1)?,
                    zh: row.get(2)?,
                    level: row.get(3)?,
                    context_sentence: row.get(4)?,
                    state: "new".to_string(),
                })
            })
            .map_err(|e| e.to_string())?;
        for row in rows {
            result.push(row.map_err(|e| e.to_string())?);
        }
    }

    Ok(result)
}

#[derive(Serialize)]
pub struct ReviewResult {
    pub next_review_at: String,
    pub scheduled_days: i64,
    pub state: String,
}

#[tauri::command]
pub fn db_review_card(
    word_id: i64,
    rating: String,
    conn: TauriState<'_, AppState>,
) -> Result<ReviewResult, String> {
    let rating = match rating.as_str() {
        "again" => Rating::Again,
        "hard" => Rating::Hard,
        "good" => Rating::Good,
        other => return Err(format!("invalid rating: {other} (expected again/hard/good)")),
    };

    let db = db::lock_db(&conn)?;

    let existing: Option<(f64, f64, i64, i64, i64, i64, String, String, i64)> = db
        .query_row(
            "SELECT stability, difficulty, elapsed_days, scheduled_days, review_count, lapses, next_review_at, last_reviewed_at, state
             FROM srs_records WHERE entity_id = ?1 AND entity_type = 'word'",
            params![word_id],
            |row| {
                Ok((
                    row.get(0)?,
                    row.get(1)?,
                    row.get(2)?,
                    row.get(3)?,
                    row.get(4)?,
                    row.get(5)?,
                    row.get::<_, Option<String>>(6)?.unwrap_or_default(),
                    row.get::<_, Option<String>>(7)?.unwrap_or_default(),
                    row.get(8)?,
                ))
            },
        )
        .optional()
        .map_err(|e| e.to_string())?;

    let now = Utc::now();
    let card = match existing {
        Some((stability, difficulty, elapsed_days, scheduled_days, reps, lapses, due_s, last_review_s, state_i)) => Card {
            due: if due_s.is_empty() { now } else { sql_to_dt(&due_s) },
            stability,
            difficulty,
            elapsed_days,
            scheduled_days,
            reps: reps as i32,
            lapses: lapses as i32,
            state: state_from_i64(state_i),
            last_review: if last_review_s.is_empty() { now } else { sql_to_dt(&last_review_s) },
        },
        None => Card::new(),
    };

    let fsrs = FSRS::default();
    let scheduling = fsrs.next(card, now, rating);
    let new_card = scheduling.card;

    let next_review_at = dt_to_sql(new_card.due);
    let last_reviewed_at = dt_to_sql(new_card.last_review);
    let state_str = state_to_str(new_card.state).to_string();

    db.execute(
        "INSERT INTO srs_records
            (entity_id, entity_type, srs_level, srs_ease, review_count, last_reviewed_at, next_review_at,
             stability, difficulty, elapsed_days, scheduled_days, lapses, state)
         VALUES (?1, 'word', ?2, 2.5, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)
         ON CONFLICT(entity_id, entity_type) DO UPDATE SET
            srs_level = ?2, review_count = ?3, last_reviewed_at = ?4, next_review_at = ?5,
            stability = ?6, difficulty = ?7, elapsed_days = ?8, scheduled_days = ?9,
            lapses = ?10, state = ?11",
        params![
            word_id,
            new_card.state as i64,
            new_card.reps as i64,
            last_reviewed_at,
            next_review_at,
            new_card.stability,
            new_card.difficulty,
            new_card.elapsed_days,
            new_card.scheduled_days,
            new_card.lapses as i64,
            new_card.state as i64,
        ],
    )
    .map_err(|e| e.to_string())?;

    db.execute(
        "INSERT INTO daily_streaks (date, quiz_done) VALUES (date('now'), 1)
         ON CONFLICT(date) DO UPDATE SET quiz_done = quiz_done + 1",
        [],
    )
    .map_err(|e| e.to_string())?;

    Ok(ReviewResult {
        next_review_at,
        scheduled_days: new_card.scheduled_days,
        state: state_str,
    })
}

File: mobile_tanwords/src/db/srs.ts
/**
 * FSRS review scheduling — port of desktop `app/core/src/db/srs.rs`,
 * rs-fsrs → ts-fsrs. Shares the `srs_records` table with the desktop app:
 * State discriminants are 0=New, 1=Learning, 2=Review, 3=Relearning in both
 * crates, and `rating` is the desktop's "again" | "hard" | "good" (numbers
 * 1..4, "easy" included, are accepted as an extension).
 */
import { fsrs, Rating, State, type Card, type Grade } from "ts-fsrs";
import { getDb } from "./connection";
import type { DueCard, ReviewResult, SrsRating, SrsState } from "@/hooks/useDB.types";

/** Words introduced per db_get_due_cards call when the reviewer isn't
 *  working through a review backlog — keeps a big vocabulary from dumping
 *  hundreds of "new" cards into one session. */
export const DEFAULT_NEW_LIMIT = 20;

function stateToStr(state: State): SrsState {
  switch (state) {
    case State.Learning:
      return "learning";
    case State.Review:
      return "review";
    case State.Relearning:
      return "relearning";
    default:
      return "new";
  }
}

function stateFromInt(v: number): State {
  switch (v) {
    case 1:
      return State.Learning;
    case 2:
      return State.Review;
    case 3:
      return State.Relearning;
    default:
      return State.New;
  }
}

function dtToSql(dt: Date): string {
  // RFC3339, like Rust's to_rfc3339() — "Z" instead of "+00:00", same instant,
  // and sqlToDt below parses both forms.
  return dt.toISOString();
}

/**
 * Port of `sql_to_dt`. Stored values are RFC3339 (written by dtToSql, desktop
 * or mobile) or SQLite's default datetime('now') format
 * ("YYYY-MM-DD HH:MM:SS", UTC without a marker) for rows never touched by
 * FSRS. Both are parsed WITHOUT leaning on the JS engine's lenient Date
 * parsing (which would read the naive form as local time); anything
 * unrecognized falls back to now, like the Rust code.
 */
export function sqlToDt(s: string): Date {
  const rfc3339 = s.match(
    /^(\d{4})-(\d{2})-(\d{2})[Tt](\d{2}):(\d{2}):(\d{2})(\.\d+)?([Zz]|[+-]\d{2}:?\d{2})$/
  );
  if (rfc3339) {
    const [, y, mo, d, h, mi, se, frac, tz] = rfc3339;
    let ms = Date.UTC(
      Number(y),
      Number(mo) - 1,
      Number(d),
      Number(h),
      Number(mi),
      Number(se),
      frac ? Math.round(Number(frac) * 1000) : 0
    );
    if (tz.toUpperCase() !== "Z") {
      const sign = tz.startsWith("-") ? -1 : 1;
      const digits = tz.replace(/[^0-9]/g, "");
      const offsetMin = sign * (Number(digits.slice(0, 2)) * 60 + Number(digits.slice(2, 4)));
      ms -= offsetMin * 60_000;
    }
    return new Date(ms);
  }
  const naive = s.match(/^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})$/);
  if (naive) {
    const [, y, mo, d, h, mi, se] = naive;
    return new Date(Date.UTC(Number(y), Number(mo) - 1, Number(d), Number(h), Number(mi), Number(se)));
  }
  return new Date();
}

function ratingFromInput(rating: SrsRating | "easy" | 1 | 2 | 3 | 4): Grade {
  switch (rating) {
    case "again":
    case 1:
      return Rating.Again;
    case "hard":
    case 2:
      return Rating.Hard;
    case "good":
    case 3:
      return Rating.Good;
    case "easy":
    case 4:
      return Rating.Easy;
    default:
      throw new Error(`invalid rating: ${rating} (expected again/hard/good)`);
  }
}

/**
 * Count of cards ready for review right now: the due backlog plus a capped
 * batch of never-reviewed words, using the same cap as db_get_due_cards'
 * default so the Dashboard badge matches what opening the reviewer shows.
 *
 * Compares dates in JS rather than SQL because next_review_at may hold either
 * RFC3339 (written by db_review_card) or SQLite's own datetime() format
 * (written by the older db_save_quiz_result path) — the two don't sort
 * correctly against each other as raw strings.
 */
export async function db_get_review_count(): Promise<number> {
  const db = getDb();
  const now = Date.now();

  const dueRows = await db.getAllAsync<{ next_review_at: string }>(
    "SELECT next_review_at FROM srs_records WHERE entity_type = 'word' AND next_review_at IS NOT NULL"
  );
  const dueCount = dueRows.filter((r) => sqlToDt(r.next_review_at).getTime() <= now).length;

  const newRow = await db.getFirstAsync<{ n: number }>(
    "SELECT COUNT(*) AS n FROM words w LEFT JOIN srs_records sr ON sr.entity_id = w.id AND sr.entity_type = 'word' WHERE sr.id IS NULL"
  );
  const newCount = newRow?.n ?? 0;

  return dueCount + Math.min(newCount, DEFAULT_NEW_LIMIT);
}

const ZH_EXPR =
  "COALESCE((SELECT wd.zh FROM word_definitions wd WHERE wd.word_id = w.id ORDER BY wd.sort_order LIMIT 1), '')";
const CONTEXT_EXPR =
  "COALESCE(" +
  "(SELECT wd.example_en FROM word_definitions wd " +
  "WHERE wd.word_id = w.id AND wd.example_en IS NOT NULL AND wd.example_en != '' " +
  "ORDER BY wd.sort_order LIMIT 1), " +
  "''" +
  ")";

export async function db_get_due_cards(args?: { newLimit?: number | null }): Promise<DueCard[]> {
  const db = getDb();
  const newLimit = args?.newLimit ?? DEFAULT_NEW_LIMIT;

  const result: DueCard[] = [];

  // Backlog: previously-scheduled reviews that are now due.
  const dueRows = await db.getAllAsync<{
    id: number;
    word: string;
    zh: string;
    level: string | null;
    context_sentence: string;
    state: number;
  }>(
    `SELECT w.id, w.word, ${ZH_EXPR} AS zh, w.level, ${CONTEXT_EXPR} AS context_sentence, sr.state
     FROM words w
     JOIN srs_records sr ON sr.entity_id = w.id AND sr.entity_type = 'word'
     WHERE sr.next_review_at <= ?
     ORDER BY sr.next_review_at ASC`,
    [dtToSql(new Date())]
  );
  for (const r of dueRows) {
    result.push({
      word_id: r.id,
      word: r.word,
      zh: r.zh,
      level: r.level,
      context_sentence: r.context_sentence,
      state: stateToStr(stateFromInt(r.state)),
    });
  }

  // New: words never reviewed, capped so a big vocabulary doesn't flood the session.
  const newRows = await db.getAllAsync<{
    id: number;
    word: string;
    zh: string;
    level: string | null;
    context_sentence: string;
  }>(
    `SELECT w.id, w.word, ${ZH_EXPR} AS zh, w.level, ${CONTEXT_EXPR} AS context_sentence
     FROM words w
     LEFT JOIN srs_records sr ON sr.entity_id = w.id AND sr.entity_type = 'word'
     WHERE sr.id IS NULL
     ORDER BY w.created_at ASC
     LIMIT ?`,
    [newLimit]
  );
  for (const r of newRows) {
    result.push({
      word_id: r.id,
      word: r.word,
      zh: r.zh,
      level: r.level,
      context_sentence: r.context_sentence,
      state: "new",
    });
  }

  return result;
}

export async function db_review_card(args: {
  wordId: number;
  rating: SrsRating | "easy" | 1 | 2 | 3 | 4;
}): Promise<ReviewResult> {
  const db = getDb();
  const rating = ratingFromInput(args.rating);

  const row = await db.getFirstAsync<{
    stability: number;
    difficulty: number;
    elapsed_days: number;
    scheduled_days: number;
    review_count: number;
    lapses: number;
    next_review_at: string | null;
    last_reviewed_at: string | null;
    state: number;
  }>(
    "SELECT stability, difficulty, elapsed_days, scheduled_days, review_count, lapses, next_review_at, last_reviewed_at, state " +
      "FROM srs_records WHERE entity_id = ? AND entity_type = 'word'",
    [args.wordId]
  );

  const now = new Date();
  const card: Card = row
    ? {
        due: row.next_review_at ? sqlToDt(row.next_review_at) : now,
        stability: row.stability,
        difficulty: row.difficulty,
        elapsed_days: row.elapsed_days,
        scheduled_days: row.scheduled_days,
        learning_steps: 0,
        reps: row.review_count,
        lapses: row.lapses,
        state: stateFromInt(row.state),
        last_review: row.last_reviewed_at ? sqlToDt(row.last_reviewed_at) : now,
      }
    : {
        // rs-fsrs Card::new(): due/last_review = now, everything else zeroed.
        due: now,
        stability: 0,
        difficulty: 0,
        elapsed_days: 0,
        scheduled_days: 0,
        learning_steps: 0,
        reps: 0,
        lapses: 0,
        state: State.New,
        last_review: now,
      };

  const f = fsrs();
  const newCard = f.repeat(card, now)[rating].card;

  const nextReviewAt = dtToSql(newCard.due);
  const lastReviewedAt = dtToSql(newCard.last_review ?? now);
  const stateStr = stateToStr(newCard.state);
  const stateInt = newCard.state as number;

  await db.runAsync(
    "INSERT INTO srs_records " +
      "(entity_id, entity_type, srs_level, srs_ease, review_count, last_reviewed_at, next_review_at, " +
      "stability, difficulty, elapsed_days, scheduled_days, lapses, state) " +
      "VALUES (?, 'word', ?, 2.5, ?, ?, ?, ?, ?, ?, ?, ?, ?) " +
      "ON CONFLICT(entity_id, entity_type) DO UPDATE SET " +
      "srs_level = ?, review_count = ?, last_reviewed_at = ?, next_review_at = ?, " +
      "stability = ?, difficulty = ?, elapsed_days = ?, scheduled_days = ?, " +
      "lapses = ?, state = ?",
    [
      args.wordId,
      stateInt,
      newCard.reps,
      lastReviewedAt,
      nextReviewAt,
      newCard.stability,
      newCard.difficulty,
      newCard.elapsed_days,
      newCard.scheduled_days,
      newCard.lapses,
      stateInt,
      stateInt,
      newCard.reps,
      lastReviewedAt,
      nextReviewAt,
      newCard.stability,
      newCard.difficulty,
      newCard.elapsed_days,
      newCard.scheduled_days,
      newCard.lapses,
      stateInt,
    ]
  );

  await db.runAsync(
    "INSERT INTO daily_streaks (date, quiz_done) VALUES (date('now'), 1) " +
      "ON CONFLICT(date) DO UPDATE SET quiz_done = quiz_done + 1"
  );

  return {
    next_review_at: nextReviewAt,
    scheduled_days: newCard.scheduled_days,
    state: stateStr,
  };
}
